src/logic/otp2/controller.ts

```diff
@@ -536,15 +525,136 @@ async function searchTaxiFrontBack(
+/**
+ * TODO 7/12-21: Clean up function signature and types once OTP1 is gone.
+ * As we have removed recursion, initialSearchDate can be replaced with
+ * searchDate. searchFilter could
+ * be kept separate from Otp2SearchParams.
+ */
+export async function searchTransit(
+    { initialSearchDate, searchFilter, ...searchParams }: Otp2SearchParams,
+    extraHeaders: { [key: string]: string },
+): Promise<TransitTripPatterns> {
+    const { searchDate, arriveBy } = searchParams
+
+    const getTripPatternsParams = {
+        ...searchParams,
+        modes: filterModesAndSubModes(searchFilter),
+    }
+
+    // We do two searches in parallel here to speed things up a bit. One is a
+    // flexible search where we explicitly look for trips that may include means
+    // of transport that has to be booked in advance the second is a regular
+    // search.
+    const [
+        flexibleResults,
+        [regularTripPatternsUnfiltered, initialMetadata, routingErrors],
+    ] = await Promise.all([
+        searchFlexible(searchParams),
+        getTripPatterns(getTripPatternsParams),
+    ])
+
+    let queries = [
+        ...flexibleResults.queries,
+        getTripPatternsQuery(getTripPatternsParams),
+    ]
+
+    let metadata = initialMetadata
+
+    let regularTripPatterns = regularTripPatternsUnfiltered
+        .filter(isValidTransitAlternative)
+        .map(replaceQuay1ForOsloSWithUnknown)
+
+    const nextSearchDateFromMetadata =
+        metadata && getNextSearchDateFromMetadata(metadata, arriveBy)
+
+    const flexibleTripPattern = flexibleResults.tripPatterns[0]
+    const hasFlexibleResultsOnly =
+        flexibleTripPattern && !regularTripPatterns.length
+    if (hasFlexibleResultsOnly) {
+        // Flexible may return results in the future that are outside the
+        // original search window. There may still exist normal transport in the
+        // time between the search window end and the first suggested flexible
+        // result, so we do a new search to try to find those. For example, if
+        // the search window ends on midnight Friday, and the first suggested
+        // flexible result is on Monday morning, we can probably still find a
+        // normal transport option on Saturday.
+        const beforeFlexibleResult = await searchBeforeFlexible(
+            nextSearchDateFromMetadata || searchDate,
+            arriveBy,
+            flexibleTripPattern,
+            getTripPatternsParams,
+        )
+        regularTripPatterns = beforeFlexibleResult.tripPatterns
+        metadata = beforeFlexibleResult.metadata
+        queries = [...queries, ...beforeFlexibleResult.queries]
+    }
+
+    // If we have any noStopsInRange errors, we couldn't find a means of
+    // transport from where the traveler wants to start or end the trip. Try to
+    // find an option using taxi for those parts instead.
+    const noStopsInRangeErrors = routingErrors.filter(
```

Comment:
I moved taxi stuff to after 'beforeFlexible' search, as beforeFlexible still tries to get the regularTripPatterns. It feels like it belongs next to the initial search. Taxi is independent and may come after.

Reply:
...or rather, it is dependent on the initial noStopsInRange, but it doesn't matter if we put it after beforeFlexible, as that would have been run anyhow.

Reply:
Buuuuut, I'm starting to think that we don't need to search a second time (using beforeFlexible) if noStopsInRange is not empty, as we would get the same result? @draperunner care to shed some light on this?

Reply:
I think you're right. If we get `noStopsInRange`, `searchBeforeFlexible` shouldn't return anything.